File: 06_binary_search/solution.py
from typing import List

def binary_search(arr: List[int], target: int)-> int:

    left: int = 0
    right: int = len(arr) - 1

    while left <= right:
        mid = left + (right - left) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] > target:
            right = mid - 1
        elif arr[mid] < target:
            left = mid + 1
        
    return -1

if __name__ == "__main__":
    # Happy path
    assert binary_search([-1, 0, 3, 5, 9, 12], 9) == 4
    
    # Target not found (simulando uma query miss)
    assert binary_search([-1, 0, 3, 5, 9, 12], 2) == -1 
    
    # Empty list edge case (cluster vazio)
    assert binary_search([], 5) == -1
    
    # Single element list
    assert binary_search([5], 5) == 0
    print("All tests passed! Production ready.")

File: 06_binary_search/solution.go
package main

// BinarySearch finds the index of a target integer in a sorted slice.
// It returns -1 if the target is not found.
func BinarySearch(arr []int, target int) int {
	left := 0
	right := len(arr) - 1

	for left <= right {
		// Calculate mid safely to avoid integer overflow
		mid := left + (right-left)/2

		if arr[mid] == target {
			return mid // Target found, route the request here
		} else if arr[mid] < target {
			left = mid + 1 // Discard the entire left partition
		} else {
			right = mid - 1 // Discard the entire right partition
		}
	}

	return -1 // Target not found in the cluster
}
